File: src/main/scheduler.ts
import { powerMonitor } from 'electron'
import type { ReminderItem, ScheduleItem } from '@shared/types'
import { applyJitter, computeIntervalAt, computeScheduleAt } from '@shared/schedule-core'

type Source = 'reminder' | 'schedule'

/**
 * 双源调度：间隔提醒（r:id）按固定周期（可带抖动），定时日程（s:id）按钟点。
 * 单定时器只对准最近的下一次触发；sync 按 key+签名 差量更新，改某项不重置其他项。
 * 时刻计算全部在 shared/schedule-core.ts（纯函数，可单测）。
 */
export class Scheduler {
  private timer: NodeJS.Timeout | null = null
  /** 每个已启用项的下一次触发时间（epoch ms） */
  private nextAt = new Map<string, number>()
  /** 变更检测签名：间隔值 / time+weekdays+date */
  private sigs = new Map<string, string>()
  private intervalsMs = new Map<string, number>()
  private scheduleItems = new Map<string, ScheduleItem>()
  /** 休眠错过提醒的策略：'fire' 唤醒后补发，'skip' 丢弃 */
  missedPolicy: 'fire' | 'skip' = 'fire'
  /** 间隔抖动比例（0-0.5），0 关闭 */
  jitterRatio = 0

  constructor(private readonly onRemind: (source: Source, itemId: string) => void) {
    powerMonitor.on('resume', this.handleResume)
  }

  sync(reminders: ReminderItem[], schedules: ScheduleItem[]): void {
    const desired = new Map<string, string>()
    for (const r of reminders) {
      if (r.enabled) desired.set(`r:${r.id}`, String(r.intervalMinutes))
    }
    for (const s of schedules) {
      if (s.enabled) desired.set(`s:${s.id}`, `${s.time}|${[...s.weekdays].sort().join(',')}|${s.date ?? ''}`)
    }

    // 移除已删除/停用/签名变化的项
    for (const key of [...this.nextAt.keys()]) {
      if (desired.get(key) !== this.sigs.get(key)) this.nextAt.delete(key)
    }
    this.sigs = new Map(desired)

    const now = Date.now()
    this.scheduleItems.clear()
    for (const s of schedules) {
      if (s.enabled) this.scheduleItems.set(s.id, s)
    }

    for (const [key, sig] of desired) {
      if (this.nextAt.has(key)) continue
      const nowDate = new Date(now)
      if (key.startsWith('r:')) {
        const minutes = Number(sig)
        this.intervalsMs.set(key, minutes * 60_000)
        this.nextAt.set(key, this.jitteredInterval(minutes, nowDate))
      } else {
        const item = this.scheduleItems.get(key.slice(2))
        if (!item) continue
        const at = computeScheduleAt(item, nowDate)
        if (at > now) this.nextAt.set(key, at)
      }
    }
    this.arm()
  }

  stop(): void {
    if (this.timer) clearTimeout(this.timer)
    this.timer = null
    this.nextAt.clear()
    this.sigs.clear()
    this.intervalsMs.clear()
    this.scheduleItems.clear()
  }

  /** 手动触发后重置间隔项的周期；定时日程按钟点触发，无需重置 */
  resetItem(itemId: string): void {
    const key = `r:${itemId}`
    const ms = this.intervalsMs.get(key)
    if (ms && this.nextAt.has(key)) {
      this.nextAt.set(key, this.jitteredInterval(ms / 60_000, new Date()))
      this.arm()
    }
  }

  /** 距最近一次提醒的分钟数（向上取整），无已启用项时返回 0 */
  nextInMinutes(): number {
    const at = this.earliest()
    if (at === null) return 0
    return Math.max(0, Math.ceil((at - Date.now()) / 60_000))
  }

  /** 最近将触发的项：kind + id，无则 null */
  nextDue(): { source: Source; id: string } | null {
    const at = this.earliest()
    if (at === null) return null
    for (const [key, t] of this.nextAt) {
      if (t === at) {
        const [source, id] = key.startsWith('r:') ? (['reminder', key.slice(2)] as const) : (['schedule', key.slice(2)] as const)
        return { source, id }
      }
    }
    return null
  }

  /** 间隔周期 + 抖动 */
  private jitteredInterval(intervalMinutes: number, now: Date): number {
    return applyJitter(computeIntervalAt(intervalMinutes, now), intervalMinutes * 60_000, this.jitterRatio)
  }

  private earliest(): number | null {
    let min: number | null = null
    for (const at of this.nextAt.values()) {
      if (min === null || at < min) min = at
    }
    return min
  }

  /** 只挂一个定时器，对准最近的下一次触发 */
  private arm(): void {
    if (this.timer) clearTimeout(this.timer)
    this.timer = null
    const at = this.earliest()
    if (at === null) return
    this.timer = setTimeout(() => this.fire(this.nextKey()!), Math.max(0, at - Date.now()))
  }

  private nextKey(): string | null {
    const at = this.earliest()
    if (at === null) return null
    for (const [key, t] of this.nextAt) {
      if (t === at) return key
    }
    return null
  }

  private fire(key: string): void {
    this.timer = null
    const now = Date.now()
    if (key.startsWith('r:')) {
      const ms = this.intervalsMs.get(key)
      if (!ms) return
      this.nextAt.set(key, this.jitteredInterval(ms / 60_000, new Date(now)))
      this.onRemind('reminder', key.slice(2))
    } else {
      const item = this.scheduleItems.get(key.slice(2))
      if (!item) return
      this.onRemind('schedule', item.id)
      if (item.date) {
        // 一次性日程：触发后不再排期，由主进程停用
        this.nextAt.delete(key)
      } else {
        const next = computeScheduleAt(item, new Date(now))
        if (next > now) this.nextAt.set(key, next)
      }
    }
    this.arm()
  }

  /** 系统唤醒后：已到期的项按策略补发或跳过，然后重新排期 */
  private handleResume = (): void => {
    if (!this.nextAt.size) return
    const now = Date.now()
    for (const [key, at] of [...this.nextAt]) {
      if (at > now) continue
      if (this.missedPolicy === 'fire') this.onRemind(key.startsWith('r:') ? 'reminder' : 'schedule', key.slice(2))
      if (key.startsWith('r:')) {
        const ms = this.intervalsMs.get(key)
        if (ms) this.nextAt.set(key, this.jitteredInterval(ms / 60_000, new Date(now)))
      } else {
        const item = this.scheduleItems.get(key.slice(2))
        const next = item && !item.date ? computeScheduleAt(item, new Date(now)) : 0
        if (next > now) this.nextAt.set(key, next)
        else this.nextAt.delete(key)
      }
    }
    this.arm()
  }
}

File: src-tauri/src/scheduler.rs
use std::collections::HashMap;
use std::sync::mpsc::{channel, Receiver, RecvTimeoutError, Sender};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use chrono::Local;
use tauri::AppHandle;

use crate::config::{ReminderItem, ScheduleItem};
use crate::schedule::{apply_jitter, compute_anchored_next, compute_schedule_at};

#[derive(Clone, Copy, PartialEq)]
pub enum Source {
    Reminder,
    Schedule,
}

pub struct SyncData {
    pub reminders: Vec<ReminderItem>,
    pub schedules: Vec<ScheduleItem>,
}

pub enum Cmd {
    Sync(SyncData),
    ResetItem(String),
    SetMissedPolicy(String),
}

pub struct SchedulerHandle {
    tx: Sender<Cmd>,
}

impl SchedulerHandle {
    pub fn sync(&self, data: SyncData) {
        let _ = self.tx.send(Cmd::Sync(data));
    }
    pub fn reset_item(&self, item_id: &str) {
        let _ = self.tx.send(Cmd::ResetItem(item_id.to_string()));
    }
    pub fn set_missed_policy(&self, policy: &str) {
        let _ = self.tx.send(Cmd::SetMissedPolicy(policy.to_string()));
    }
}

fn now_ms() -> i64 {
    Local::now().timestamp_millis()
}

/// 双源调度（移植自 Electron 版 scheduler.ts）：
/// 单线程循环只对准最近一次触发，按 key+签名差量更新。
pub fn spawn(
    app: AppHandle,
    mirror: Arc<Mutex<HashMap<String, i64>>>,
    on_fire: impl Fn(Source, &str) + Send + 'static,
) -> SchedulerHandle {
    let (tx, rx) = channel::<Cmd>();
    std::thread::spawn(move || run_loop(app, rx, mirror, on_fire));
    SchedulerHandle { tx }
}

struct State {
    next_at: HashMap<String, i64>,
    sigs: HashMap<String, String>,
    intervals_ms: HashMap<String, i64>,
    anchors: HashMap<String, i64>,
    schedule_items: HashMap<String, ScheduleItem>,
    jitter_ratio: f64,
    /// 休眠错过提醒的策略：'fire' 唤醒后补发，'skip' 丢弃
    missed_policy: String,
}

impl State {
    fn new() -> Self {
        State {
            next_at: HashMap::new(),
            sigs: HashMap::new(),
            intervals_ms: HashMap::new(),
            anchors: HashMap::new(),
            schedule_items: HashMap::new(),
            jitter_ratio: 0.0,
            missed_policy: "fire".into(),
        }
    }

    fn next_interval_at(&self, interval_ms: i64, anchor_at: i64, now: i64) -> i64 {
        if anchor_at > 0 {
            compute_anchored_next(anchor_at, interval_ms, now)
        } else {
            apply_jitter(now + interval_ms, interval_ms, self.jitter_ratio)
        }
    }

    fn sync(&mut self, data: SyncData) {
        let mut desired: HashMap<String, String> = HashMap::new();
        for r in &data.reminders {
            if r.enabled {
                desired.insert(
                    format!("r:{}", r.id),
                    format!(
                        "{}|{}",
                        r.interval_seconds,
                        r.anchor_at.map(|a| a.to_string()).unwrap_or_default()
                    ),
                );
            }
        }
        for s in &data.schedules {
            if s.enabled {
                let mut wds = s.weekdays.clone();
                wds.sort();
                let wd_str = wds
                    .iter()
                    .map(|d| d.to_string())
                    .collect::<Vec<_>>()
                    .join(",");
                desired.insert(
                    format!("s:{}", s.id),
                    format!(
                        "{}|{}|{}",
                        s.time,
                        wd_str,
                        s.date.clone().unwrap_or_default()
                    ),
                );
            }
        }

        self.next_at
            .retain(|k, _| desired.get(k) == self.sigs.get(k));
        self.sigs = desired.clone();
        self.schedule_items.clear();
        self.anchors.clear();

        for s in &data.schedules {
            if s.enabled {
                self.schedule_items.insert(s.id.clone(), s.clone());
            }
        }
        for r in &data.reminders {
            if r.enabled {
                if let Some(a) = r.anchor_at {
                    self.anchors.insert(format!("r:{}", r.id), a as i64);
                }
            }
        }

        let now = now_ms();
        for r in &data.reminders {
            if !r.enabled {
                continue;
            }
            let key = format!("r:{}", r.id);
            if self.next_at.contains_key(&key) {
                continue;
            }
            let ms = r.interval_seconds as i64 * 1000;
            self.intervals_ms.insert(key.clone(), ms);
            let anchor = r.anchor_at.map(|a| a as i64).unwrap_or(0);
            self.next_at
                .insert(key, self.next_interval_at(ms, anchor, now));
        }
        for s in &data.schedules {
            if !s.enabled {
                continue;
            }
            let key = format!("s:{}", s.id);
            if self.next_at.contains_key(&key) {
                continue;
            }
            let at = compute_schedule_at(&s.time, &s.weekdays, s.date.as_deref(), Local::now());
            if at > now {
                self.next_at.insert(key, at);
            }
        }
    }

    fn earliest(&self) -> Option<i64> {
        self.next_at.values().min().copied()
    }

    fn earliest_key(&self) -> Option<String> {
        let at = self.earliest()?;
        self.next_at
            .iter()
            .find(|(_, t)| **t == at)
            .map(|(k, _)| k.clone())
    }

    fn fire(&mut self, key: &str) -> Option<(Source, String, i64)> {
        let fired_at = now_ms();
        if let Some(id) = key.strip_prefix("r:") {
            let ms = *self.intervals_ms.get(key)?;
            let anchor = self.anchors.get(key).copied().unwrap_or(0);
            let prev_planned = self.next_at.get(key).copied().unwrap_or(fired_at);
            let next = if anchor > 0 {
                compute_anchored_next(anchor, ms, prev_planned.min(fired_at))
            } else {
                self.next_interval_at(ms, 0, fired_at)
            };
            self.next_at.insert(key.to_string(), next);
            Some((Source::Reminder, id.to_string(), fired_at - prev_planned))
        } else if let Some(id) = key.strip_prefix("s:") {
            let item = self.schedule_items.get(id)?.clone();
            let prev_planned = self.next_at.get(key).copied().unwrap_or(fired_at);
            if item.date.is_some() {
                self.next_at.remove(key);
            } else {
                let next = compute_schedule_at(&item.time, &item.weekdays, None, Local::now());
                if next > fired_at {
                    self.next_at.insert(key.to_string(), next);
                } else {
                    self.next_at.remove(key);
                }
            }
            Some((Source::Schedule, id.to_string(), fired_at - prev_planned))
        } else {
            None
        }
    }

    fn reset_item(&mut self, item_id: &str) {
        let key = format!("r:{item_id}");
        if let (Some(ms), true) = (
            self.intervals_ms.get(&key).copied(),
            self.next_at.contains_key(&key),
        ) {
            let anchor = self.anchors.get(&key).copied().unwrap_or(0);
            self.next_at
                .insert(key, self.next_interval_at(ms, anchor, now_ms()));
        }
    }
}

fn run_loop(
    _app: AppHandle,
    rx: Receiver<Cmd>,
    mirror: Arc<Mutex<HashMap<String, i64>>>,
    on_fire: impl Fn(Source, &str) + Send,
) {
    let mut st = State::new();
    let apply_cmd = |st: &mut State, cmd: Cmd| match cmd {
        Cmd::Sync(data) => st.sync(data),
        Cmd::ResetItem(id) => st.reset_item(&id),
        Cmd::SetMissedPolicy(p) => st.missed_policy = p,
    };
    let push_mirror = |st: &State| {
        *mirror.lock().unwrap() = st.next_at.clone();
    };
    // skip 策略：休眠等导致迟到超过 90s 的触发丢弃（仍照常重排）
    let should_fire = |st: &State, overdue: i64| !(st.missed_policy == "skip" && overdue > 90_000);
    loop {
        let mut changed = false;
        while let Ok(cmd) = rx.try_recv() {
            apply_cmd(&mut st, cmd);
            changed = true;
        }
        if changed {
            push_mirror(&st);
        }
        let now = now_ms();
        match st.earliest() {
            None => match rx.recv() {
                Ok(cmd) => {
                    apply_cmd(&mut st, cmd);
                    push_mirror(&st);
                }
                Err(_) => break,
            },
            Some(at) if at <= now => {
                if let Some(key) = st.earliest_key() {
                    if let Some((source, id, overdue)) = st.fire(&key) {
                        push_mirror(&st);
                        if should_fire(&st, overdue) {
                            on_fire(source, &id);
                        }
                    }
                }
            }
            Some(at) => {
                let wait = Duration::from_millis((at - now) as u64);
                match rx.recv_timeout(wait) {
                    Ok(cmd) => {
                        apply_cmd(&mut st, cmd);
                        push_mirror(&st);
                    }
                    Err(RecvTimeoutError::Timeout) => {
                        if let Some(key) = st.earliest_key() {
                            if let Some((source, id, overdue)) = st.fire(&key) {
                                push_mirror(&st);
                                if should_fire(&st, overdue) {
                                    on_fire(source, &id);
                                }
                            }
                        }
                    }
                    Err(RecvTimeoutError::Disconnected) => break,
                }
            }
        }
    }
}
